File: experimental/attention/quant_block_sparse_attn/torch_ops_extension/custom_ops/csrc/npu_quant_block_sparse_attn_metadata.cpp
/**
 * Copyright (c) 2026 Huawei Technologies Co., Ltd.
 * This program is free software, you can redistribute it and/or modify it under the terms and conditions of
 * CANN Open Software License Agreement Version 2.0 (the "License").
 * Please refer to the License for details. You may not use this file except in compliance with the License.
 * THIS SOFTWARE IS PROVIDED ON AN "AS IS" BASIS, WITHOUT WARRANTIES OF ANY KIND, EITHER EXPRESS OR IMPLIED,
 * INCLUDING BUT NOT LIMITED TO NON-INFRINGEMENT, MERCHANTABILITY, OR FITNESS FOR A PARTICULAR PURPOSE.
 * See LICENSE in the root of the software repository for the full text of the License.
 */

#include <torch/library.h>
#include "ops_common.h"

namespace custom {
using namespace at_npu::native;

constexpr int64_t QBSA_METADATA_OUTPUT_SIZE = 2048;

c10::optional<at::Tensor> get_qbsa_valid_tensor(
    const c10::optional<at::Tensor> &tensorOpt, const at::Device &device)
{
    if (tensorOpt.has_value() && tensorOpt.value().defined()) {
        return tensorOpt;
    }
    return c10::optional<at::Tensor>(torch::empty({0}, torch::dtype(torch::kInt32).device(device)));
}

at::Tensor npu_quant_block_sparse_attn_metadata_npu(
    const at::Tensor &sparseSeqLen, int64_t numHeadsQ, int64_t numHeadsKv, int64_t headDim,
    const c10::optional<at::Tensor> &cuSeqlensQ, const c10::optional<at::Tensor> &cuSeqlensKv,
    const c10::optional<at::Tensor> &sequsedQ, const c10::optional<at::Tensor> &sequsedKv,
    int64_t batchSize, int64_t sparseBlockSizeQ, int64_t sparseBlockSizeK, int64_t quantMode, int64_t maskMode,
    int64_t maxSeqlenQ, int64_t maxSeqlenKv, c10::string_view layoutQ, c10::string_view layoutKv,
    c10::string_view layoutSparseIndices)
{
    at::Device outputDevice = sparseSeqLen.device();
    at::Tensor output = torch::empty({QBSA_METADATA_OUTPUT_SIZE}, torch::dtype(torch::kInt32).device(outputDevice));

    auto cuSeqlensQValue = get_qbsa_valid_tensor(cuSeqlensQ, outputDevice);
    auto cuSeqlensKvValue = get_qbsa_valid_tensor(cuSeqlensKv, outputDevice);
    auto sequsedQValue = get_qbsa_valid_tensor(sequsedQ, outputDevice);
    auto sequsedKvValue = get_qbsa_valid_tensor(sequsedKv, outputDevice);

    std::string layoutQStr = std::string(layoutQ);
    std::string layoutKvStr = std::string(layoutKv);
    std::string layoutSparseIndicesStr = std::string(layoutSparseIndices);
    char *layoutQPtr = const_cast<char *>(layoutQStr.c_str());
    char *layoutKvPtr = const_cast<char *>(layoutKvStr.c_str());
    char *layoutSparseIndicesPtr = const_cast<char *>(layoutSparseIndicesStr.c_str());

    EXEC_NPU_CMD_V1(aclnnQuantBlockSparseAttnMetadata,
        sparseSeqLen, cuSeqlensQValue, cuSeqlensKvValue, sequsedQValue, sequsedKvValue,
        batchSize, numHeadsQ, numHeadsKv, headDim, sparseBlockSizeQ, sparseBlockSizeK, quantMode, maskMode,
        maxSeqlenQ, maxSeqlenKv, layoutQPtr, layoutKvPtr, layoutSparseIndicesPtr, output);
    return output;
}

at::Tensor npu_quant_block_sparse_attn_metadata_meta(
    const at::Tensor &sparseSeqLen, int64_t numHeadsQ, int64_t numHeadsKv, int64_t headDim,
    const c10::optional<at::Tensor> &cuSeqlensQ, const c10::optional<at::Tensor> &cuSeqlensKv,
    const c10::optional<at::Tensor> &sequsedQ, const c10::optional<at::Tensor> &sequsedKv,
    int64_t batchSize, int64_t sparseBlockSizeQ, int64_t sparseBlockSizeK, int64_t quantMode, int64_t maskMode,
    int64_t maxSeqlenQ, int64_t maxSeqlenKv, c10::string_view layoutQ, c10::string_view layoutKv,
    c10::string_view layoutSparseIndices)
{
    return torch::empty({QBSA_METADATA_OUTPUT_SIZE}, sparseSeqLen.options().dtype(torch::kInt32));
}
} // namespace custom

TORCH_LIBRARY_IMPL(custom, PrivateUse1, m) {
    m.impl("npu_quant_block_sparse_attn_metadata", &custom::npu_quant_block_sparse_attn_metadata_npu);
}

TORCH_LIBRARY_IMPL(custom, Meta, m) {
    m.impl("npu_quant_block_sparse_attn_metadata", &custom::npu_quant_block_sparse_attn_metadata_meta);
}

File: experimental/attention/quant_block_sparse_attn/torch_ops_extension/custom_ops/converter/npu_quant_block_sparse_attn_metadata.py
# Copyright (c) 2026 Huawei Technologies Co., Ltd.
# This program is free software, you can redistribute it and/or modify it under the terms and conditions of
# CANN Open Software License Agreement Version 2.0 (the "License").
# Please refer to the License for details. You may not use this file except in compliance with the License.
# THIS SOFTWARE IS PROVIDED ON AN "AS IS" BASIS, WITHOUT WARRANTIES OF ANY KIND, EITHER EXPRESS OR IMPLIED,
# INCLUDING BUT NOT LIMITED TO NON-INFRINGEMENT, MERCHANTABILITY, OR FITNESS FOR A PARTICULAR PURPOSE.
# See LICENSE in the root of the software repository for the full text of the License.

from typing import List, Optional

import torch
import torchair
from torchair._ge_concrete_graph.fx2ge_converter import register_fx_node_ge_converter
from torchair.ge._ge_graph import Tensor, TensorSpec
from torchair.ge import attr


@register_fx_node_ge_converter(torch.ops.custom.npu_quant_block_sparse_attn_metadata.default)
def convert_npu_quant_block_sparse_attn_metadata(
    sparse_seq_len: Tensor,
    num_heads_q: int,
    num_heads_kv: int,
    head_dim: int,
    *,
    cu_seqlens_q: Optional[Tensor] = None,
    cu_seqlens_kv: Optional[Tensor] = None,
    seqused_q: Optional[Tensor] = None,
    seqused_kv: Optional[Tensor] = None,
    batch_size: int = 0,
    sparse_block_size_q: int = 128,
    sparse_block_size_k: int = 128,
    quant_mode: int = 1,
    mask_mode: int = 3,
    max_seqlen_q: int = 0,
    max_seqlen_kv: int = 0,
    layout_q: str = "TND",
    layout_kv: str = "PA_BNSD",
    layout_sparse_indices: str = "B_N_Qb_Kb",
    meta_outputs: List[TensorSpec] = None,
):
    stream_info = torch.npu.get_stream_limit(torch.npu.current_stream())
    aic_core_num = stream_info.get("cube_core_num")
    aiv_core_num = stream_info.get("vector_core_num")
    soc_version = torch.npu.get_device_properties().name
    return torchair.ge.custom_op(
        "QuantBlockSparseAttnMetadata",
        inputs={
            "sparse_seq_len": sparse_seq_len,
            "cu_seqlens_q": cu_seqlens_q,
            "cu_seqlens_kv": cu_seqlens_kv,
            "seqused_q": seqused_q,
            "seqused_kv": seqused_kv,
        },
        attrs={
            "batch_size": attr.Int(batch_size),
            "num_heads_q": attr.Int(num_heads_q),
            "num_heads_kv": attr.Int(num_heads_kv),
            "head_dim": attr.Int(head_dim),
            "sparse_block_size_q": attr.Int(sparse_block_size_q),
            "sparse_block_size_k": attr.Int(sparse_block_size_k),
            "quant_mode": attr.Int(quant_mode),
            "mask_mode": attr.Int(mask_mode),
            "max_seqlen_q": attr.Int(max_seqlen_q),
            "max_seqlen_kv": attr.Int(max_seqlen_kv),
            "layout_q": attr.Str(layout_q),
            "layout_kv": attr.Str(layout_kv),
            "layout_sparse_indices": attr.Str(layout_sparse_indices),
            "soc_version": attr.Str(soc_version),
            "aic_core_num": attr.Int(aic_core_num),
            "aiv_core_num": attr.Int(aiv_core_num),
        },
        outputs=["metadata"],
    )
